### PR TITLE
Add Lua support

### PR DESCRIPTION
**Pull request by SGauvin**
Add lua support!!
- Just create a `LuaEffect` with the filename and the correct functions and it works
- Lua scripts MUST include `framework.lua`
- All lua scripts need to do is to populate the `Colors` table, which is the right size already
- Lua scripts must also create their schema to receive settings which they can useAdd lua support!!
- Just create a `LuaEffect` with the filename and the correct functions and it works
- Lua scripts MUST include `framework.lua`
- All lua scripts need to do is to populate the `Colors` table, which is the right size already
- Lua scripts must also create their schema to receive settings which they can use

### CODE REVIEW

File: src/resources/effects/lua.rs [hunk 1]
```diff
@@ -104,6 +104,7 @@ impl LuaEffect {
         // });
         let dst = self.lua_leds_buffer.as_mut_ptr() as *mut u8;
         let src = data.as_ptr();
+        // Safety: Perfs
```
Comment (DIGarnier) [inline src/resources/effects/lua.rs]: queringe

File: src/resources/effects/lua.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        &mut self,
+        leds: &mut [Color],
+        settings: &LuaEffectSettings,
+    ) -> Result<(), LuaEffectRuntimeError> {
+        self.lua
+            .globals()
+            .set("settings", self.lua.to_value(&settings.settings).unwrap())
+            .map_err(LuaEffectRuntimeError::Lua)?;
+
+        let resize_fn: Function = match self.lua.globals().get("Resize_Colors") {
```
Comment (DIGarnier) [inline src/resources/effects/lua.rs]: map_err ?

File: src/resources/effects/lua.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            Ok(resize_fn) => Ok(resize_fn),
+            Err(_) => Err(LuaEffectRuntimeError::MissingFrameworkImport),
+        }?;
+        resize_fn
+            .call::<_, Value>(leds.len())
+            .map_err(LuaEffectRuntimeError::Lua)?;
+
+        let tick_fn: Function = match self.lua.globals().get("Tick") {
```
Comment (DIGarnier) [inline src/resources/effects/lua.rs]: map_err?

File: src/resources/effects/lua.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            Ok(tick_fn) => Ok(tick_fn),
+            Err(_) => Err(LuaEffectRuntimeError::MissingTickFunction),
+        }?;
+
+        tick_fn
+            .call::<_, ()>(())
+            .map_err(LuaEffectRuntimeError::Lua)?;
+
+        let set_colors_fn: Function = match self.lua.globals().get("Set_colors") {
```
Comment (DIGarnier) [inline src/resources/effects/lua.rs]: map_err

File: scripts/framework.lua [hunk 5]
```diff
@@ -0,0 +1,19 @@
+Colors = {}
+
+function Set_colors()
```
Comment (DIGarnier) [inline scripts/framework.lua:3]: tu pourrais get plus de perf ici je crois

https://www.dannyguo.com/blog/how-to-concatenate-strings-in-lua/

File: src/resources/effects/lua.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+#[derive(Debug)]
+pub enum LuaEffectRuntimeError {
+    Lua(Error),
+    WrongColorsLen,
+    MissingTickFunction,
+    MissingFrameworkImport,
+}
+
+#[allow(unused)]
```
Comment (samichoulo911) [inline src/resources/effects/lua.rs:27]: Do we still need this?

Comment (SGauvin) [reply src/resources/effects/lua.rs:27]: Ouais a cause des json schema
Va falloir check comment on veut manage ca

File: scripts/framework.lua [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	local data = {}
+	for i, value in pairs(Colors) do
+		local index = ((i - 1) * 3)
+		data[index + 1] = string.char(value.r)
```
Comment (DIGarnier) [inline scripts/framework.lua:7]: j'pense que tu es pas obligé de string.char? je sais pas si ça change les perfs.

File: src/resources/effects/lua.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+#[derive(Clone, Debug)]
+pub struct LuaEffectSettings {
+    pub settings: serde_json::Value,
+}
+
+impl LuaEffect {
+    pub fn new(filename: &str) -> Result<Self, LuaEffectLoadError> {
+        let (lua, json_schema, compiled_json_schema) = Self::get_lua_effect(filename)?;
+        Ok(LuaEffect {
```
Comment (DIGarnier) [inline src/resources/effects/lua.rs]:
```suggestion
        Ok(Self {
```

File: scripts/framework.lua (hunk 7 above)
Comment (SGauvin) [reply scripts/framework.lua:7]: Ouinnn pcq sinon ca va litteralement ajouter le chiffre a ton string.
Genre 11 va append "11" au string

Comment (DIGarnier) [reply scripts/framework.lua:7]: ah kk ouan lul

File: src/resources/effects/lua.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        &mut self,
+        leds: &mut [Color],
+        settings: &LuaEffectSettings,
+    ) -> Result<(), LuaEffectRuntimeError> {
+        self.lua
+            .globals()
+            .set("settings", self.lua.to_value(&settings.settings).unwrap())
+            .map_err(LuaEffectRuntimeError::Lua)?;
+
+        let resize_fn: Function = self
```
Comment (samichoulo911) [inline src/resources/effects/lua.rs:60]: Pourquoi on doit resize a chaque tick?

**Review by samichoulo911 — APPROVED** [review]

**Pull request merged**